src/engine/bufferscalers/rubberbandwrapper.cpp

```diff
@@ -0,0 +1,174 @@
+    if (m_pInstances.size() == 1) {
+        return m_pInstances[0]->retrieve(output, samples);
+    } else {
+        // ensure we don't fetch more samples than we really have available.
+        samples = std::min(static_cast<size_t>(available()), samples);
+        for (const auto& stretcher : m_pInstances) {
+#ifdef MIXXX_DEBUG_ASSERTIONS_ENABLED
+            size_t numSamplesRequested =
+#endif
+                    stretcher->retrieve(output, samples);
+            // there is something very wrong if we got a different of amount
+            // of samples than we requested
+            DEBUG_ASSERT(numSamplesRequested == samples);
+            output += getChannelPerWorker();
+        }
+        return samples;
+    }
+}
+size_t RubberBandWrapper::getInputIncrement() const {
+    VERIFY_OR_DEBUG_ASSERT(isValid()) {
+        return -1;
+    }
+    return m_pInstances[0]->getInputIncrement();
+}
+size_t RubberBandWrapper::getLatency() const {
+    VERIFY_OR_DEBUG_ASSERT(isValid()) {
+        return -1;
+    }
+    return m_pInstances[0]->getLatency();
+}
+double RubberBandWrapper::getPitchScale() const {
+    VERIFY_OR_DEBUG_ASSERT(isValid()) {
+        return -1;
+    }
+    return m_pInstances[0]->getPitchScale();
+}
+// See
+// https://github.com/breakfastquay/rubberband/commit/72654b04ea4f0707e214377515119e933efbdd6c
+// for how these two functions were implemented within librubberband itself
+size_t RubberBandWrapper::getPreferredStartPad() const {
+    VERIFY_OR_DEBUG_ASSERT(isValid()) {
+        return -1;
+    }
+#if RUBBERBANDV3
+    return m_pInstances[0]->getPreferredStartPad();
+#else
+    // `getPreferredStartPad()` returns `window_size / 2`, while with
+    // `getLatency()` both time stretching engines return `window_size / 2 /
+    // pitch_scale`
+    return static_cast<size_t>(std::ceil(
+            m_pInstances[0]->getLatency() * m_pInstances[0]->getPitchScale()));
+#endif
+}
+size_t RubberBandWrapper::getStartDelay() const {
+    VERIFY_OR_DEBUG_ASSERT(isValid()) {
+        return -1;
+    }
+#if RUBBERBANDV3
+    return m_pInstances[0]->getStartDelay();
+#else
+    // In newer Rubber Band versions `getLatency()` is a deprecated alias for
+    // `getStartDelay()`, so they should behave the same. In the commit linked
+    // above the behavior was different for the R3 stretcher, but that was only
+    // during the initial betas of Rubberband 3.0 so we shouldn't have to worry
+    // about that.
+    return m_pInstances[0]->getLatency();
+#endif
+}
+void RubberBandWrapper::process(const float* const* input, size_t samples, bool isFinal) {
+    if (m_pInstances.size() == 1) {
+        return m_pInstances[0]->process(input, samples, isFinal);
+    } else {
+        RubberBandWorkerPool* pPool = RubberBandWorkerPool::instance();
+        for (auto& pInstance : m_pInstances) {
+            pInstance->set(input, samples, isFinal);
+            if (!pPool->tryStart(pInstance.get())) {
```

Comment:
No being able to start is part of the valid expectations. This happens if there is no more worker available and indicate that the caller should take care of performing the stretching itself (This is why on the follow line, we `run` the task directly on the following line). This is to reduce the chance of main thread to wait, since it will perform active stretching on the last instance (the pool is always sized with at most N-1 for that reason), so chances are that it will be the last one terminating the process, and it won't have to wait on the condition in `waitReady`